# WeRateDogs Twitter Analysis

This report analyzes the Twitter archive of the user [@dog_rates](https://twitter.com/dog_rates), also known as WeRateDogs. The Twitter user rates other peoples' dogs and provides humorous content about them.

The Twitter archive data is being used here to practice data wrangling techniques to create analyses and visualizations about the data.

## Import Libraries

In [1]:
import pandas as pd
import numpy as np
import requests
import tweepy
import json

# Data Wrangling

## Gather Data

In [2]:
# Load Twitter archive CSV into dataframe
tweets = pd.read_csv('twitter-archive-enhanced.csv')

In [27]:
# Download image predictions file from Udacity and save
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
with open(url.split('/')[-1], mode='wb') as file:
    file.write(response.content)

In [3]:
# Load image predictions file into dataframe
images = pd.read_csv('image-predictions.tsv', sep='\t')

In [28]:
# Connect to Twitter API
consumer_key = 'YOUR CONSUMER KEY'
consumer_secret = 'YOUR CONSUMER SECRET'
access_token = 'YOUR ACCESS TOKEN'
access_secret = 'YOUR ACCESS SECRET'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

# Query Twitter API for additional tweet data
with open('tweet_json.txt', mode='a') as f:
    for tweet_id in tweets.tweet_id:
        try:
            print(tweet_id)
            tweet = api.get_status(tweet_id, tweet_mode='extended')._json
            json.dump(tweet, f)
            f.write('\n')
        except Exception as e:
            print(str(tweet_id) + ': Error accessing tweet - the tweet may have been deleted.')

8.92421E+17
8.92421E+17: Error accessing tweet - the tweet may have been deleted.
8.92177E+17
8.92177E+17: Error accessing tweet - the tweet may have been deleted.
8.91815E+17
8.91815E+17: Error accessing tweet - the tweet may have been deleted.
8.9169E+17
8.9169E+17: Error accessing tweet - the tweet may have been deleted.
8.91328E+17
8.91328E+17: Error accessing tweet - the tweet may have been deleted.
8.91088E+17
8.91088E+17: Error accessing tweet - the tweet may have been deleted.
8.90972E+17
8.90972E+17: Error accessing tweet - the tweet may have been deleted.
8.90729E+17
8.90729E+17: Error accessing tweet - the tweet may have been deleted.
8.90609E+17
8.90609E+17: Error accessing tweet - the tweet may have been deleted.
8.9024E+17
8.9024E+17: Error accessing tweet - the tweet may have been deleted.
8.90007E+17
8.90007E+17: Error accessing tweet - the tweet may have been deleted.
8.89881E+17
8.89881E+17: Error accessing tweet - the tweet may have been deleted.
8.89665E+17
8.89665E

KeyboardInterrupt: 

In [4]:
# Create list of dictionaries to build line by line, then convert to dataframe
df_list = []
with open('tweet_json.txt', mode='r', encoding='utf-8') as f:
    for line in f:
        l = json.loads(line)
        tweet_id = l['id_str']
        retweet_count = l['retweet_count']
        favorite_count = l['favorite_count']
        
        df_list.append({'tweet_id': tweet_id,
                       'retweet_count': retweet_count,
                       'favorite_count': favorite_count})

tweet_stats = pd.DataFrame(df_list, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

## Assess Data

In [5]:
# Display full columns entries (non-truncated)
pd.set_option('display.max_colwidth', -1)

In [6]:
tweets

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193 NaN                    NaN                    
1     892177421306343426 NaN                    NaN                    
2     891815181378084864 NaN                    NaN                    
3     891689557279858688 NaN                    NaN                    
4     891327558926688256 NaN                    NaN                    
5     891087950875897856 NaN                    NaN                    
6     890971913173991426 NaN                    NaN                    
7     890729181411237888 NaN                    NaN                    
8     890609185150312448 NaN                    NaN                    
9     890240255349198849 NaN                    NaN                    
10    890006608113172480 NaN                    NaN                    
11    889880896479866881 NaN                    NaN                    
12    889665388333682689 NaN                    NaN                    
13    889638837579907072 NaN                    NaN                    
14    889531135344209921 NaN                    NaN                    
15    889278841981685760 NaN                    NaN                    
16    888917238123831296 NaN                    NaN                    
17    888804989199671297 NaN                    NaN                    
18    888554962724278272 NaN                    NaN                    
19    888202515573088257 NaN                    NaN                    
20    888078434458587136 NaN                    NaN                    
21    887705289381826560 NaN                    NaN                    
22    887517139158093824 NaN                    NaN                    
23    887473957103951883 NaN                    NaN                    
24    887343217045368832 NaN                    NaN                    
25    887101392804085760 NaN                    NaN                    
26    886983233522544640 NaN                    NaN                    
27    886736880519319552 NaN                    NaN                    
28    886680336477933568 NaN                    NaN                    
29    886366144734445568 NaN                    NaN                    
...                  ...  ..                     ..                    
2326  666411507551481857 NaN                    NaN                    
2327  666407126856765440 NaN                    NaN                    
2328  666396247373291520 NaN                    NaN                    
2329  666373753744588802 NaN                    NaN                    
2330  666362758909284353 NaN                    NaN                    
2331  666353288456101888 NaN                    NaN                    
2332  666345417576210432 NaN                    NaN                    
2333  666337882303524864 NaN                    NaN                    
2334  666293911632134144 NaN                    NaN                    
2335  666287406224695296 NaN                    NaN                    
2336  666273097616637952 NaN                    NaN                    
2337  666268910803644416 NaN                    NaN                    
2338  666104133288665088 NaN                    NaN                    
2339  666102155909144576 NaN                    NaN                    
2340  666099513787052032 NaN                    NaN                    
2341  666094000022159362 NaN                    NaN                    
2342  666082916733198337 NaN                    NaN                    
2343  666073100786774016 NaN                    NaN                    
2344  666071193221509120 NaN                    NaN                    
2345  666063827256086533 NaN                    NaN                    
2346  666058600524156928 NaN                    NaN                    
2347  666057090499244032 NaN                    NaN                    
2348  666055525042405380 NaN                    NaN                    
2349  666051853826850816 NaN                    NaN     

In [7]:
images

tweet_id  \
0     666020888022790149   
1     666029285002620928   
2     666033412701032449   
3     666044226329800704   
4     666049248165822465   
5     666050758794694657   
6     666051853826850816   
7     666055525042405380   
8     666057090499244032   
9     666058600524156928   
10    666063827256086533   
11    666071193221509120   
12    666073100786774016   
13    666082916733198337   
14    666094000022159362   
15    666099513787052032   
16    666102155909144576   
17    666104133288665088   
18    666268910803644416   
19    666273097616637952   
20    666287406224695296   
21    666293911632134144   
22    666337882303524864   
23    666345417576210432   
24    666353288456101888   
25    666362758909284353   
26    666373753744588802   
27    666396247373291520   
28    666407126856765440   
29    666411507551481857   
...                  ...   
2045  886366144734445568   
2046  886680336477933568   
2047  886736880519319552   
2048  886983233522544640   
2049  887101392804085760   
2050  887343217045368832   
2051  887473957103951883   
2052  887517139158093824   
2053  887705289381826560   
2054  888078434458587136   
2055  888202515573088257   
2056  888554962724278272   
2057  888804989199671297   
2058  888917238123831296   
2059  889278841981685760   
2060  889531135344209921   
2061  889638837579907072   
2062  889665388333682689   
2063  889880896479866881   
2064  890006608113172480   
2065  890240255349198849   
2066  890609185150312448   
2067  890729181411237888   
2068  890971913173991426   
2069  891087950875897856   
2070  891327558926688256   
2071  891689557279858688   
2072  891815181378084864   
2073  892177421306343426   
2074  892420643555336193   

                                                                                      jpg_url  \
0     https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg                                           
1     https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg                                           
2     https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg                                           
3     https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg                                           
4     https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg                                           
5     https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg                                           
6     https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg                                           
7     https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg                                           
8     https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg                                           
9     https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg                                           
10    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg                                           
11    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg                                           
12    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg                                           
13    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg                                           
14    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg                                           
15    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg                                           
16    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg                                           
17    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg                                           
18    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg                                           
19    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg                                           
20    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg                                           
21    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg                                           
22    https://pbs.twimg.com/media/CT9OwFIWEAMuRj

In [8]:
tweet_stats

tweet_id  retweet_count  favorite_count
0     892420643555336193  8558           38696         
1     892177421306343426  6291           33164         
2     891815181378084864  4174           24963         
3     891689557279858688  8681           42071         
4     891327558926688256  9451           40225         
5     891087950875897856  3125           20173         
6     890971913173991426  2082           11817         
7     890729181411237888  18984          65365         
8     890609185150312448  4281           27723         
9     890240255349198849  7447           31864         
10    890006608113172480  7365           30582         
11    889880896479866881  4990           27727         
12    889665388333682689  10110          48014         
13    889638837579907072  4566           27111         
14    889531135344209921  2243           15063         
15    889278841981685760  5448           25244         
16    888917238123831296  4515           29005         
17    888804989199671297  4363           25531         
18    888554962724278272  3595           19857         
19    888078434458587136  3510           21703         
20    887705289381826560  5415           30107         
21    887517139158093824  11718          46127         
22    887473957103951883  18310          68947         
23    887343217045368832  10449          33594         
24    887101392804085760  5988           30455         
25    886983233522544640  7808           35071         
26    886736880519319552  3312           12039         
27    886680336477933568  4488           22366         
28    886366144734445568  3208           21146         
29    886267009285017600  4              116           
...                  ... ..              ...           
2315  666411507551481857  328            448           
2316  666407126856765440  41             110           
2317  666396247373291520  86             166           
2318  666373753744588802  93             189           
2319  666362758909284353  574            779           
2320  666353288456101888  73             221           
2321  666345417576210432  139            299           
2322  666337882303524864  92             199           
2323  666293911632134144  357            509           
2324  666287406224695296  66             149           
2325  666273097616637952  76             175           
2326  666268910803644416  35             104           
2327  666104133288665088  6637           14351         
2328  666102155909144576  13             80            
2329  666099513787052032  68             156           
2330  666094000022159362  74             164           
2331  666082916733198337  45             119           
2332  666073100786774016  164            322           
2333  666071193221509120  62             148           
2334  666063827256086533  220            475           
2335  666058600524156928  57             112           
2336  666057090499244032  142            298           
2337  666055525042405380  252            434           
2338  666051853826850816  853            1225          
2339  666050758794694657  58             132           
2340  666049248165822465  41             109           
2341  666044226329800704  141            299           
2342  666033412701032449  45             125           
2343  666029285002620928  47             129           
2344  666020888022790149  517            2563          

[2345 rows x 3 columns]

In [34]:
# Look at tweets with lowercase names
tweets[tweets.name.str.islower()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
22    887517139158093824 NaN                    NaN                    
56    881536004380872706 NaN                    NaN                    
118   869988702071779329 NaN                    NaN                    
169   859196978902773760 NaN                    NaN                    
193   855459453768019968 NaN                    NaN                    
335   832645525019123713 NaN                    NaN                    
369   828650029636317184 NaN                    NaN                    
542   806219024703037440 NaN                    NaN                    
649   792913359805018113 NaN                    NaN                    
682   788552643979468800 NaN                    NaN                    
759   778396591732486144 NaN                    NaN                    
773   776249906839351296 NaN                    NaN                    
801   772581559778025472 NaN                    NaN                    
819   770655142660169732 NaN                    NaN                    
822   770093767776997377 NaN                    NaN                    
852   765395769549590528 NaN                    NaN                    
924   755206590534418437 NaN                    NaN                    
988   748977405889503236 NaN                    NaN                    
992   748692773788876800 NaN                    NaN                    
993   748575535303884801 NaN                    NaN                    
1002  747885874273214464 NaN                    NaN                    
1004  747816857231626240 NaN                    NaN                    
1017  746872823977771008 NaN                    NaN                    
1025  746369468511756288 NaN                    NaN                    
1031  745422732645535745 NaN                    NaN                    
1040  744223424764059648 NaN                    NaN                    
1049  743222593470234624 NaN                    NaN                    
1063  741067306818797568 NaN                    NaN                    
1071  740214038584557568 NaN                    NaN                    
1095  736392552031657984 NaN                    NaN                    
...                  ...  ..                     ..                    
2191  668955713004314625 NaN                    NaN                    
2198  668815180734689280 NaN                    NaN                    
2204  668636665813057536 NaN                    NaN                    
2211  668614819948453888 NaN                    NaN                    
2212  668587383441514497 NaN                    NaN                    
2218  668507509523615744 NaN                    NaN                    
2222  668466899341221888 NaN                    NaN                    
2235  668171859951755264 NaN                    NaN                    
2249  667861340749471744 NaN                    NaN                    
2255  667773195014021121 NaN                    NaN                    
2264  667538891197542400 NaN                    NaN                    
2273  667470559035432960 NaN                    NaN                    
2287  667177989038297088 NaN                    NaN                    
2304  666983947667116034 NaN                    NaN                    
2311  666781792255496192 NaN                    NaN                    
2314  666701168228331520 NaN                    NaN                    
2326  666411507551481857 NaN                    NaN                    
2327  666407126856765440 NaN                    NaN                    
2333  666337882303524864 NaN                    NaN                    
2334  666293911632134144 NaN                    NaN                    
2335  666287406224695296 NaN                    NaN                    
2345  666063827256086533 NaN                    NaN                    
2346  666058600524156928 NaN                    NaN                    
2347  666057090499244032 NaN                    NaN     

In [35]:
# Look at tweets with decimal ratings
decimal_rating = tweets.text.str.contains('\d{1,2}\.\d+/\d+')
tweets[decimal_rating]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
45    883482846933004288 NaN                    NaN                    
340   832215909146226688 NaN                    NaN                    
695   786709082849828864 NaN                    NaN                    
763   778027034220126208 NaN                    NaN                    
1689  681340665377193984  6.813394e+17           4.196984e+09          
1712  680494726643068929 NaN                    NaN                    

                      timestamp  \
45    2017-07-08 00:28:19 +0000   
340   2017-02-16 13:11:49 +0000   
695   2016-10-13 23:23:56 +0000   
763   2016-09-20 00:24:34 +0000   
1689  2015-12-28 05:07:27 +0000   
1712  2015-12-25 21:06:00 +0000   

                                                                                  source  \
45    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
340   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
695   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
763   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1689  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1712  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                   text  \
45    This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948                            
340   RT @dog_rates: This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wu…                        
695   This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS                                    
763   This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq   
1689  I've been told there's a slight possibility he's checking his mirror. We'll bump to 9.5/10. Still a menace                                                          
1712  Here we have uncovered an entire battalion of holiday puppers. Average of 11.26/10 https://t.co/eNm2S6p9BD                                                          

      retweeted_status_id  retweeted_status_user_id  \
45   NaN                  NaN                         
340   7.867091e+17         4.196984e+09               
695  NaN                  NaN                         
763  NaN                  NaN                         
1689 NaN                  NaN                         
1712 NaN                  NaN                         

     retweeted_status_timestamp  \
45    NaN                         
340   2016-10-13 23:23:56 +0000   
695   NaN                         
763   NaN                         
1689  NaN                         
1712  NaN                         

                                                                                                                        expanded_urls  \
45    https://twitter.com/dog_rates/status/883482846933004288/photo/1,https://twitter.com/dog_rates/status/883482846933004288/photo/1   
340   https://twitter.com/dog_rates/status/786709082849828864/photo/1                                                                   
695   https://twitter.com/dog_rates/status/786709082849828864/photo/1                                                                   
763   https://twitter.com/dog_rates/status/778027034220126208/photo/1                                                                   
1689  NaN                                                                                                

In [36]:
tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [6]:
images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [7]:
tweet_stats.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2345 entries, 0 to 2344
Data columns (total 3 columns):
tweet_id          2345 non-null object
retweet_count     2345 non-null int64
favorite_count    2345 non-null int64
dtypes: int64(2), object(1)
memory usage: 55.0+ KB


In [37]:
tweets.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03  7.800000e+01           7.800000e+01          
mean   7.427716e+17  7.455079e+17           2.014171e+16          
std    6.856705e+16  7.582492e+16           1.252797e+17          
min    6.660209e+17  6.658147e+17           1.185634e+07          
25%    6.783989e+17  6.757419e+17           3.086374e+08          
50%    7.196279e+17  7.038708e+17           4.196984e+09          
75%    7.993373e+17  8.257804e+17           4.196984e+09          
max    8.924206e+17  8.862664e+17           8.405479e+17          

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count  1.810000e+02         1.810000e+02              2356.000000        
mean   7.720400e+17         1.241698e+16              13.126486          
std    6.236928e+16         9.599254e+16              45.876648          
min    6.661041e+17         7.832140e+05              0.000000           
25%    7.186315e+17         4.196984e+09              10.000000          
50%    7.804657e+17         4.196984e+09              11.000000          
75%    8.203146e+17         4.196984e+09              12.000000          
max    8.874740e+17         7.874618e+17              1776.000000        

       rating_denominator  
count  2356.000000         
mean   10.455433           
std    6.745237            
min    0.000000            
25%    10.000000           
50%    10.000000           
75%    10.000000           
max    170.000000

All of the rankings I've seen on WeRateDogs have had a denominator of 10. Above we can see that the min and max of the rating denominators are 0 and 170. Let's look closer at rating denominators not equal to 10.

In [38]:
# Tweets with rating denominators not equal to 10
tweets[tweets.rating_denominator != 10]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640  8.352460e+17           2.625958e+07          
342   832088576586297345  8.320875e+17           3.058208e+07          
433   820690176645140481 NaN                    NaN                    
516   810984652412424192 NaN                    NaN                    
784   775096608509886464 NaN                    NaN                    
902   758467244762497024 NaN                    NaN                    
1068  740373189193256964 NaN                    NaN                    
1120  731156023742988288 NaN                    NaN                    
1165  722974582966214656 NaN                    NaN                    
1202  716439118184652801 NaN                    NaN                    
1228  713900603437621249 NaN                    NaN                    
1254  710658690886586372 NaN                    NaN                    
1274  709198395643068416 NaN                    NaN                    
1351  704054845121142784 NaN                    NaN                    
1433  697463031882764288 NaN                    NaN                    
1598  686035780142297088  6.860340e+17           4.196984e+09          
1634  684225744407494656  6.842229e+17           4.196984e+09          
1635  684222868335505415 NaN                    NaN                    
1662  682962037429899265 NaN                    NaN                    
1663  682808988178739200  6.827884e+17           4.196984e+09          
1779  677716515794329600 NaN                    NaN                    
1843  675853064436391936 NaN                    NaN                    
2335  666287406224695296 NaN                    NaN                    

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
342   2017-02-16 04:45:50 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
784   2016-09-11 22:20:06 +0000   
902   2016-07-28 01:00:57 +0000   
1068  2016-06-08 02:41:38 +0000   
1120  2016-05-13 16:15:54 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1598  2016-01-10 04:04:10 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1662  2016-01-01 16:30:13 +0000   
1663  2016-01-01 06:22:03 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                                                  source  \
313   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
342   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
433   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
516   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
784   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
902   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1068  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1120  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1165  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1202  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1228  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1254  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1274  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1351  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1433  <a href="http://twitter.com/download/iphone" rel="nofollow">

In [39]:
tweets[tweets.rating_denominator == 11].tweet_id

784     775096608509886464
1068    740373189193256964
1662    682962037429899265
Name: tweet_id, dtype: int64

### Quality
`tweets`:
- Some of the tweets are retweets
- There is no image in some of the tweets
- Some names are recorded as 'None' when the text format is 'named xxxx' and 'name is xxxx'
- Name of dog in tweet_id '740373189193256964' is named 'Bretagne' not 'None'
- The lowercase names are not actually names
- Ratings with decimals are interpreted incorrectly (e.g. 5/10 instead of 13.5/10)
- Many fractions interpreted as ratings instead of actual rating that comes later in tweet (960/00, 11/15/15, 24/7, 9/11, 4/20, 50/50, 7/11, 3 1/2)
- Missing names and missing dog stages represented as 'None'
- Erroneous data types for tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, retweeted_status_id, retweeted_user_id, retweeted_status_timestamp, dog stages

`images`:
- Erroneous data type for tweet_id


### Tidiness
- The variable 'dog_stage' in `tweets` is broken into 4 columns: 'doggo', 'floofer', 'pupper', and 'puppo'
- Retweet count and favorite count should be a part of `tweets` table

## Clean Data

In [9]:
# Make copies of dataframes
tweets_clean = tweets.copy()
images_clean = images.copy()
tweet_stats_clean = tweet_stats.copy()

__Some of the tweets are retweets__

### Define
Remove retweets from `tweets` table by checking if retweeted status id column is null, then remove retweet-related columns

### Code and Test

In [10]:
# Keep only tweets that have retweeted status id of null, then reset index
tweets_clean = tweets_clean[tweets_clean.retweeted_status_id.isnull()]
tweets_clean = tweets_clean.reset_index(drop=True)

In [11]:
# Check if any retweet columns have data
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2175 entries, 0 to 2174
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: float64(4), int64(3), object(1

In [12]:
# Remove retweet columns
retweet_cols = ['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp']
tweets_clean = tweets_clean.drop(retweet_cols, axis=1)

In [13]:
# Check columns are removed
tweets_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193 NaN                    NaN                    
1  892177421306343426 NaN                    NaN                    
2  891815181378084864 NaN                    NaN                    
3  891689557279858688 NaN                    NaN                    
4  891327558926688256 NaN                    NaN                    

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    
3  This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ                                                              
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

                                                                                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1                                                                   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1                                                                   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1                                                                   
3  https://twitter.com/dog_rates/status/891689557279858688/photo/1                                                                   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name doggo floofer pupper puppo  
0  13                10                  Phineas   None  None    None   None  
1  13                10                  Tilly     None  None    None   None  
2  12                10                  Archie    None  None    None   None  
3  13                10                  Darla     None  None    None   None  
4  12                10                  Franklin  None  None    None   None

__There is no image in some of the tweets__

### Define
Remove tweets with no image by removing tweets with null in expanded url column

### Code

In [14]:
# Remove image-less tweets and then reset index
tweets_clean = tweets_clean[tweets_clean.expanded_urls.notnull()]
tweets_clean = tweets_clean.reset_index(drop=True)

### Test

In [15]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2117 entries, 0 to 2116
Data columns (total 14 columns):
tweet_id                 2117 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                2117 non-null object
source                   2117 non-null object
text                     2117 non-null object
expanded_urls            2117 non-null object
rating_numerator         2117 non-null int64
rating_denominator       2117 non-null int64
name                     2117 non-null object
doggo                    2117 non-null object
floofer                  2117 non-null object
pupper                   2117 non-null object
puppo                    2117 non-null object
dtypes: float64(2), int64(3), object(9)
memory usage: 231.6+ KB


__The variable 'dog_stage' in `tweets` is broken into 4 columns: 'doggo', 'floofer', 'pupper', and 'puppo'__

### Define
Extract dog stage from tweet text and create dog stage column with results. Then, delete old dog stage columns ('doggo', 'floofer', 'pupper', and 'puppo').

For tweets that mention a dog stage more than once, the first mention of dog stage will be documented in dog_stage. Whether or not a dog stage was mentioned more than once (even if it was the same stage) will be documented as True or False in additional_dog_stage. The rationale is that mentioning a dog stage more than once, even if it is the same stage, emphasizes the 'cuteness' of the dog staging more than one mention.

### Code

In [16]:
# Create pandas Series of lists with dog stages mentioned in tweet text
dog_stages = pd.Series(tweets_clean.text.str.findall(r'([Dd]oggo|[Ff]loofer|[Pp]upper|[Pp]uppo)'))

# Create dog stage column with first mention of dog stage in tweet
tweets_clean['dog_stage'] = dog_stages.str[0].str.lower()

# Create additional dog stage column
tweets_clean['additional_dog_stage'] = dog_stages.str.len() > 1

In [17]:
# Delete old dog stage columns
dog_stage_cols = ['doggo', 'floofer', 'pupper', 'puppo']
tweets_clean = tweets_clean.drop(dog_stage_cols, axis=1)

### Test

In [18]:
# Confirm old dog stage columns removed and new ones added
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2117 entries, 0 to 2116
Data columns (total 12 columns):
tweet_id                 2117 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                2117 non-null object
source                   2117 non-null object
text                     2117 non-null object
expanded_urls            2117 non-null object
rating_numerator         2117 non-null int64
rating_denominator       2117 non-null int64
name                     2117 non-null object
dog_stage                373 non-null object
additional_dog_stage     2117 non-null bool
dtypes: bool(1), float64(2), int64(3), object(6)
memory usage: 184.1+ KB


In [19]:
# Check a sample of tweets to make sure dog stage column values created correctly
tweets_clean.sample(25)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1069  708738143638450176 NaN                    NaN                    
2051  667174963120574464 NaN                    NaN                    
1252  694183373896572928 NaN                    NaN                    
1580  676916996760600576 NaN                    NaN                    
1581  676897532954456065 NaN                    NaN                    
341   820690176645140481 NaN                    NaN                    
1358  687312378585812992 NaN                    NaN                    
1409  684241637099323392 NaN                    NaN                    
1221  697255105972801536 NaN                    NaN                    
2034  667491009379606528 NaN                    NaN                    
1384  685663452032069632 NaN                    NaN                    
2050  667176164155375616 NaN                    NaN                    
589   777885040357281792 NaN                    NaN                    
137   860276583193509888 NaN                    NaN                    
187   847962785489326080 NaN                    NaN                    
915   729838605770891264  7.291135e+17           4.196984e+09          
1276  692535307825213440 NaN                    NaN                    
1058  709409458133323776 NaN                    NaN                    
272   831939777352105988 NaN                    NaN                    
2081  666437273139982337 NaN                    NaN                    
2061  667062181243039745 NaN                    NaN                    
1216  697516214579523584 NaN                    NaN                    
1319  689659372465688576 NaN                    NaN                    
1169  700796979434098688 NaN                    NaN                    
1547  678255464182861824 NaN                    NaN                    

                      timestamp  \
1069  2016-03-12 19:35:15 +0000   
2051  2015-11-19 02:58:01 +0000   
1252  2016-02-01 15:39:48 +0000   
1580  2015-12-16 00:09:23 +0000   
1581  2015-12-15 22:52:02 +0000   
341   2017-01-15 17:52:40 +0000   
1358  2016-01-13 16:36:55 +0000   
1409  2016-01-05 05:14:53 +0000   
1221  2016-02-10 03:05:46 +0000   
2034  2015-11-19 23:53:52 +0000   
1384  2016-01-09 03:24:40 +0000   
2050  2015-11-19 03:02:47 +0000   
589   2016-09-19 15:00:20 +0000   
137   2017-05-04 23:34:55 +0000   
187   2017-04-01 00:04:17 +0000   
915   2016-05-10 01:00:58 +0000   
1276  2016-01-28 02:30:58 +0000   
1058  2016-03-14 16:02:49 +0000   
272   2017-02-15 18:54:34 +0000   
2081  2015-11-17 02:06:42 +0000   
2061  2015-11-18 19:29:52 +0000   
1216  2016-02-10 20:23:19 +0000   
1319  2016-01-20 04:03:02 +0000   
1169  2016-02-19 21:39:54 +0000   
1547  2015-12-19 16:47:58 +0000   

                                                                                  source  \
1069  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2051  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1252  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1580  <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                      
1581  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
341   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1358  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1409  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1221  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2034  <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                   
1384  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2050  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
589   <a href="http://twitter.com/do

__Retweet count and favorite count should be a part of `tweets` table__

### Define
Merge `tweet_stats_clean` table with `tweets_clean` table on tweet_id

### Code

In [20]:
# Change tweet id data type to string in tweets_clean so data types match
tweets_clean.tweet_id = tweets_clean.tweet_id.astype(str)

# Merge tables
tweets_clean = tweets_clean.merge(tweet_stats_clean, how='left', on='tweet_id')

### Test

In [21]:
# Check columns merged into tweets_clean
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2116
Data columns (total 14 columns):
tweet_id                 2117 non-null object
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                2117 non-null object
source                   2117 non-null object
text                     2117 non-null object
expanded_urls            2117 non-null object
rating_numerator         2117 non-null int64
rating_denominator       2117 non-null int64
name                     2117 non-null object
dog_stage                373 non-null object
additional_dog_stage     2117 non-null bool
retweet_count            2117 non-null int64
favorite_count           2117 non-null int64
dtypes: bool(1), float64(2), int64(4), object(7)
memory usage: 233.6+ KB


In [25]:
tweets_clean[['retweet_count', 'favorite_count']].sample(5)

retweet_count  favorite_count
433   2263           6707          
1709  316            1078          
222   1078           13081         
116   14908          53390         
1732  1363           2614

__Some names are recorded as 'None' when the text format is 'named xxxx' and 'name is xxxx'__

### Define
Search tweet text for 'named xxxx' and 'name is xxxx' using RegEx and capture names to record in name column

### Code

In [26]:
# Create RegEx to search for 'named' or 'name is' and extract names into name column
named_name_is = r'(?:named |name is )([A-Z][A-Za-z]+)'
tweets_clean.name = tweets_clean.text.str.extract(named_name_is, expand=False)

### Test

In [27]:
# Check rows that contain 'named' or 'name is'
tweets_clean[tweets_clean.text.str.contains(named_name_is)]

C:\Users\abarn\Anaconda3\lib\site-packages\ipykernel\__main__.py:2: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  from ipykernel import kernelapp as app


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
32    885518971528720385 NaN                    NaN                    
141   859607811541651456 NaN                    NaN                    
659   765395769549590528 NaN                    NaN                    
1453  682047327939461121 NaN                    NaN                    
1508  679736210798047232 NaN                    NaN                    
1624  675706639471788032 NaN                    NaN                    
1722  673636718965334016 NaN                    NaN                    
1801  671743150407421952 NaN                    NaN                    
1832  671147085991960577 NaN                    NaN                    
1882  670427002554466305 NaN                    NaN                    
1891  670361874861563904 NaN                    NaN                    
1894  670303360680108032 NaN                    NaN                    
1912  669923323644657664 NaN                    NaN                    
1926  669564461267722241 NaN                    NaN                    
1931  669363888236994561 NaN                    NaN                    
1955  668955713004314625 NaN                    NaN                    
1968  668636665813057536 NaN                    NaN                    
1982  668507509523615744 NaN                    NaN                    
1991  668268907921326080 NaN                    NaN                    
1999  668171859951755264 NaN                    NaN                    
2013  667861340749471744 NaN                    NaN                    
2019  667773195014021121 NaN                    NaN                    
2026  667538891197542400 NaN                    NaN                    
2029  667524857454854144 NaN                    NaN                    
2031  667509364010450944 NaN                    NaN                    
2035  667470559035432960 NaN                    NaN                    
2049  667177989038297088 NaN                    NaN                    
2065  666983947667116034 NaN                    NaN                    
2072  666781792255496192 NaN                    NaN                    
2074  666739327293083650 NaN                    NaN                    
2075  666701168228331520 NaN                    NaN                    

                      timestamp  \
32    2017-07-13 15:19:09 +0000   
141   2017-05-03 03:17:27 +0000   
659   2016-08-16 03:52:26 +0000   
1453  2015-12-30 03:55:29 +0000   
1508  2015-12-23 18:51:56 +0000   
1624  2015-12-12 15:59:51 +0000   
1722  2015-12-06 22:54:44 +0000   
1801  2015-12-01 17:30:22 +0000   
1832  2015-11-30 02:01:49 +0000   
1882  2015-11-28 02:20:27 +0000   
1891  2015-11-27 22:01:40 +0000   
1894  2015-11-27 18:09:09 +0000   
1912  2015-11-26 16:59:01 +0000   
1926  2015-11-25 17:13:02 +0000   
1931  2015-11-25 03:56:01 +0000   
1955  2015-11-24 00:54:05 +0000   
1968  2015-11-23 03:46:18 +0000   
1982  2015-11-22 19:13:05 +0000   
1991  2015-11-22 03:24:58 +0000   
1999  2015-11-21 20:59:20 +0000   
2013  2015-11-21 00:25:26 +0000   
2019  2015-11-20 18:35:10 +0000   
2026  2015-11-20 03:04:08 +0000   
2029  2015-11-20 02:08:22 +0000   
2031  2015-11-20 01:06:48 +0000   
2035  2015-11-19 22:32:36 +0000   
2049  2015-11-19 03:10:02 +0000   
2065  2015-11-18 14:18:59 +0000   
2072  2015-11-18 00:55:42 +0000   
2074  2015-11-17 22:06:57 +0000   
2075  2015-11-17 19:35:19 +0000   

                                                                                  source  \
32    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
141   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
659   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1453  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1508  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1624  <a href="http://twitter.com/down

__The lowercase names are not actually names__

### Define
Remove lowercase names from name column in `tweets_clean` table

### Code

__Name of dog in tweet_id '740373189193256964' is named 'Bretagne' not 'None'__

__Missing names and missing dog stages represented as 'None'__

### Define
Correct name column for 'Bretagne' and change missing names from 'None' to NaN (dog stages already taken care of above)

### Code

__Ratings with decimals are interpreted incorrectly (e.g. 5/10 instead of 13.5/10)__

### Define
Search tweet text for decimal ratings using RegEx and record correct rating

### Code

__Many fractions interpreted as ratings instead of actual rating that comes later in tweet (960/00, 11/15/15, 24/7, 9/11, 4/20, 50/50, 7/11, 3 1/2)__

### Define


### Code

__Erroneous data types for tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, retweeted_status_id, retweeted_user_id, retweeted_status_timestamp, dog stages__

### Define
- Change tweet_id to string in `images_clean` table (already taken care of in `tweets_clean` table)
- Change in_reply_to columns to string
- Change timestamp column to datetime
- Retweeted columns have been dropped so they are taken care of
- Change dog stage column to category data type

### Code